### ISSUE TITLE
RFC: ember-ticketfly-fields spec v2

### ISSUE THREAD

**Issue by ksin** [Ticketfly-UI/ember-ticketfly-fields#3]
After some thought, decided to dump this PR https://github.com/Ticketfly-UI/ember-ticketfly-fields/pull/2 and give the spec another shot. Here is an attempt to trim down as much as possible to minimally satisfy the design specs while making components that we can drop into an app right from install.

**tf-input-text**

Extends `Ember.TextField`. Makes it easier for someone coming in to adopt. However, I don't like that it's using two-way binding by default and the native `<input>` is pretty much the same thing. This component is basically a bundle of styles related to different states (for now?).

<img width="335" alt="screen shot 2017-02-14 at 4 44 51 pm" src="https://cloud.githubusercontent.com/assets/6626989/22956120/ee2bc728-f2d4-11e6-98e1-d0dc1782c2ee.png">

Properties it receives is `sizeGroup` which can be `large`, `small`, or `default`. 

<img width="334" alt="screen shot 2017-02-14 at 4 44 34 pm" src="https://cloud.githubusercontent.com/assets/6626989/22956115/e5078d3a-f2d4-11e6-99f1-c5a5dbe6b5c9.png">

Validations will need a bit more thinking, but for the initial implementation, user can pass in `validation=` `"success"` or `"error"`, which will simply color the input border differently. Perhaps allowing an `errorMessage` would be good too.

<img width="366" alt="screen shot 2017-02-14 at 4 45 10 pm" src="https://cloud.githubusercontent.com/assets/6626989/22956128/f8fef7ba-f2d4-11e6-988b-f976babdc2d0.png">

Normal states like hover, active, and disabled will be styled accordingly. 

`attributeBindings` it should receive include `aria-label`, `aria-describedby`, `role`, `for`.

**tf-label**

I think the label should be separate from the input component. The label shouldn't even be a component, in my opinion. But, the more we make it a ticketfly-css and native html concern, the less "drop and play" we get out of the component. Something to ponder. Regardless, we should figure a way to encourage using labels with the input (for screenreader accessibility). 

**tf-textarea**

Very similar to `tf-input-text`

@spencer516  @melissaroman @BrianSipple

**Comment by spencer516** [comment]
I'm a little confused why we're going back on this? I like the idea of not needing to worry about any accessibility concerns with all of our forms and fields — it's just "free" (<— and, we could maybe require certain properties when it's not free). But, as far as I know, the only way to do that would be to make a text field component that encapsulates all of that functionality. 

If we take this step back, then it's not clear that we're getting a ton of added value from having the components at all. Thoughts?

**Comment by ksin** [comment]
I think the thought to go back on this came from when I was considering the API for such a component.

Imagine a `{{tf-input-text}}` component that renders something like

```html
<div class="c-tf-input-text">
  <label for="some-guid" class="c-tf-input-text__label"></label>
  <input id="some-guid" class="c-tf-input-text__input">
</div>
```

The params for the input ID can be simply `inputId=`. We can generate the guids in the component if none is passed in. Great, some default accessibility already. Things like `ariaLabel`, `ariaDescribedby`, `inputRole` can also be passed through to the `<input>` as well. Nice, now we can add more accessibility attributes.

OK, but we know labels are [very important](https://www.w3.org/WAI/tutorials/forms/labels/#using-aria-label) for form controls.

How do we specify what's the input value vs what's in the label for the params? 

Some ideas that would work:
```hbs
{{tf-input-text "This is the input value" label="This is the label value"}}
{{tf-input-text value="This is the input value" label="This is the label value"}}
{{#tf-input-text value="This is the input value"}}
 This is the label value
{{/tf-input-text}}
```

I think the minimally expected use case for this component will be very similar to the Ember `{{input}}`, where one can just do: 

```hbs
{{tf-input-text "This is the input value?"}}
```

And it "just works" and is accessible, right? What that might render to:

```html
<div class="c-tf-input-text">
  <label for="some-guid" class="c-tf-input-text__label"></label>
  <input id="some-guid" class="c-tf-input-text__input" value="This is the input value?">
</div>
```

Oh no! What is the default label value if the user only entered an input value? Is the label hidden visually by default? 

There will certainly be times when the user don't want to show a label. We can perhaps pass in a `hideLabel` param so that the label is hidden visually, but still readable by screen readers. But at that point, the user probably isn't even considering what the label value is. So maybe it makes more sense to have a `showLabel` param only for when the user wants to use it with a shown label.

For example:

```hbs
{{tf-input-text "Hi there" label="Only screenreaders can see this"}}

{{tf-input-text "Lol" label="Labely labely" showLabel=true}}
```

Resulting HTML:
```html
<div class="c-tf-input-text">
  <label for="some-guid" class="c-tf-input-text__label u-hidden-visually">Only screenreaders can see this</label>
  <input id="some-guid" class="c-tf-input-text__input" value="Hi there">
</div>

<div class="c-tf-input-text">
  <label for="some-guid-2" class="c-tf-input-text__label">Labely labely</label>
  <input id="some-guid-2" class="c-tf-input-text__input" value="Lol">
</div>
```

Regardless, the user needs to be responsible for adding a label value for accessibility reasons and that means being aware of this API. There's no way to implement a default label value. If we are OK with this, then sure, I can go ahead of implement it and I'll be happy to throw in lots of API documentation. My intuition though, is to say, since there's no way to make this really accessible by default or to assume a label value, let's just have a separate label and input component.

@BrianSipple @spencer516 @melissaroman

**Comment by spencer516** [comment]
Ah — so, if we go back to the original thing I had proposed (I swear to god I don't remember writing this‽!?) — the solution I had in mind was to use contextual components so that the consumer could opt-in to composing exactly the pieces they want. Something like:

```hbs
{{#tf-input-group as |group|}}
  {{group.label "First Name"}}
  {{group.input}}
  {{group.description "You must provide a First Name"}}
{{/tf-input-group}}
```

So, if you wanted to not include a label, you simple remove the label:

```hbs
{{#tf-input-group as |group|}}
  {{group.input}}
  {{group.description "You must provide a First Name"}}
{{/tf-input-group}}

```

**Comment by ksin** [comment]
Yeah, it's the same issue though. Labels are always required if we want to encourage as much accessibility as possible. It's just a matter of whether they are hidden visually and what text goes in there.

**Comment by spencer516** [comment]
```hbs
{{#tf-input-group as |group|}}
  {{group.label hidden=true}}
  {{group.input}}
  {{group.description "You must provide a First Name"}}
{{/tf-input-group}}
```

?

**Comment by ksin** [comment]
Are you also suggesting we only use this in block-form?

**Comment by spencer516** [comment]
No — but, we should agree on what the default non-block form is. Like, perhaps we don't support a description block or hidden labels in the non-block form — only a label string and a value. 

```hbs
{{tf-input-group value="hello" label="Labely Label"}}
```

...would be equivalent to...

```hbs
{{#tf-input-group as |group|}}
  {{group.label "Labely Label"}}
  {{group.input value="hello"}}
{{/tf-input-group}}
```

...and...

```hbs
{{tf-input-group value="hello"}}
```

...would be equivalent to...

```hbs
{{#tf-input-group as |group|}}
  {{group.input value="hello"}}
{{/tf-input-group}}
```

**Comment by ksin** [comment]
Ok I can work with this :)

Lastly, on naming, I think input-text is more precise:

```
{{#tf-input-text as |group|}}
...
{{/tf-input-text}}
```

**Comment by spencer516** [comment]
^^ 👍
